packages/cli/src/commands/proxy/utils.ts

```diff
@@ -0,0 +1,84 @@
+import { conditional, trySafe } from '@silverhand/essentials';
+import chalk from 'chalk';
+import { createProxyMiddleware, responseInterceptor } from 'http-proxy-middleware';
+import { type OnProxyEvent } from 'http-proxy-middleware/dist/types.js';
+
+import { consoleLog } from '../../utils.js';
+
+import { type ProxyResponseHandler } from './types.js';
+
+export const createProxy = (targetUrl: string, onProxyResponse?: OnProxyEvent['proxyRes']) => {
+  const hasResponseHandler = Boolean(onProxyResponse);
+  return createProxyMiddleware({
+    target: targetUrl,
+    changeOrigin: true,
+    selfHandleResponse: hasResponseHandler,
+    ...conditional(
+      hasResponseHandler && {
+        on: {
+          proxyRes: onProxyResponse,
+        },
+      }
+    ),
+  });
+};
+
+/**
+ * Intercept the response from Logto endpoint and replace Logto endpoint URLs in the response with the
+ * proxy URL. The string replace happens in the following cases:
+ * - The response is a redirect response, and the response header contains `location` property.
+ * - The response contains JSON data that consists of properties such as `**_endpoint` and `redirectTo`.
+ * - The response is HTML content (logoutsource) that contains a form action URL.
+ * Note: the `issuer` and `jwks_uri` properties in the `/oidc/.well-known` response should not be replaced,
+ * even they also contain the Logto endpoint URL.
+ */
+export const createLogtoResponseHandler = async ({
+  proxyResponse,
+  request,
+  response,
+  logtoEndpointUrl,
+  proxyUrl,
+}: ProxyResponseHandler) => {
+  const { location } = proxyResponse.headers;
+  if (location) {
+    // eslint-disable-next-line @silverhand/fp/no-mutation
+    proxyResponse.headers.location = location.replace(logtoEndpointUrl.href, proxyUrl.href);
+  }
+
+  void responseInterceptor(async (responseBuffer, proxyResponse) => {
+    const responseBody = responseBuffer.toString();
+    consoleLog.info(`Response received: ${chalk.green(responseBody)}`);
+
+    if (proxyResponse.headers['content-type']?.includes('text/html')) {
+      return responseBody.replace(
+        `action="${logtoEndpointUrl.href}oidc/session/end/confirm"`,
+        `action="${proxyUrl.href}oidc/session/end/confirm"`
+      );
+    }
+
+    // eslint-disable-next-line no-restricted-syntax
+    const jsonData = trySafe(() => JSON.parse(responseBody) as Record<string, unknown>);
```

Comment:
should we also use the `content-type` header for this? thus no need for `trySafe`

Reply:
I tried this yesterday, and it seems sometimes when browser cache a GET request, the response content-type will be empty. Will try again later to confirm

Reply:
I think even if I checked for content-type, the try-catch or trySafe is inevitable, as I still have to parse a string value into JSON, then there will be potential parse error, right?

Reply:
yes, error should be omitted. however the content-type check is still good to add

Reply:
OK, updated